### PR TITLE
[#396] Add ability to select variable duration

### LINKED ISSUE #396: Add random timer range as a feature of timers
### Context
There is a use case for creating a timer that has a variable duration. For example, if you wanted to randomly space events during a game.

### Proposal
Add the ability to specify a range (min and max) of time and allow the timer to randomly choose a duration within that range.

### PR DESCRIPTION
**Pull request by malitherl**
===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/main/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

==================

Closes #396

## Changes:

-This would set a random interval after receiving a range  of numbers

### CODE REVIEW

**Review by eonarheim — CHANGES_REQUESTED** [review]
@malitherl Thank you very much for the contribution!  

I want a couple of changes before accepting:
* Implement a new random interval for each invocation of the timer
* Add the ability to pass a `Random` instance into `TimerOptions`
* Update the tests to test the above

**Comment by eonarheim** [comment]
@malitherl Let me know if you need any assistance with the requested changes 👍

**Comment by malitherl** [comment]
> @malitherl Let me know if you need any assistance with the requested changes 👍

Hi there! I have made some changes to allow the user to pass ex.Random into TimerOptions, continue to produce random intervals after reset, but I am having some difficulty with allowing the user to change the interval on reset() and also produce Random intervals

**Comment by eonarheim** [comment]
Hi @malitherl 

> Hi there! I have made some changes to allow the user to pass ex.Random into TimerOptions, continue to produce random intervals after reset, but I am having some difficulty with allowing the user to change the interval on reset() and also produce Random intervals. Any insight would be incredibly helpful, thank you

Is the issue that upon `reset(<newInterval>)` the random intervals return back to the `orignalInterval + randomRange`? 

If so I think it's because the original `interval` is closed over in this arrow function.

```typescript
this.on(()=> this.interval = interval + this.random.integer(this.randomRange[0], this.randomRange[1]))
//                              ^ JavaScript will hold onto this value from the time the () => {} was created
```

One way around that is to create a new member variable to story the "base interval" set by the constructor or the `reset()`

```typescript
private _baseInterval = 10;
private _generateRandomInterval = () => {
  return this._baseInterval + this.random.integer(this.randomRange[0], this.randomRange[1])
}

//...
// in constructor
this._callbacks = [];
this._baseInterval = this.interval = interval;
if (!!randomRange){
  if(randomRange[0] > randomRange[1]){
    throw new Error('min value must be lower than max value for range');
  }
  //We use the instance of ex.Random to generate the range
  this.random = random ?? new Random();
  this.randomRange = randomRange;

  this.interval = this._generateRandomInterval();
  this.on(() => {
    this.interval = this._generateRandomInterval();
  });
}
```

Then your `reset()` can alter the `this._baseInterval = newInterval;` 

That seems to work 👍 when I gave it a shot

image

File: src/engine/Timer.ts [hunk 1]
```diff
@@ -127,7 +159,7 @@ export class Timer {
    */
   public reset(newInterval?: number, newNumberOfRepeats?: number) {
     if (!!newInterval && newInterval >= 0) {
-      this.interval = newInterval;
+        this.interval = newInterval;
```
Comment (eonarheim) [inline src/engine/Timer.ts:162]:
```suggestion
        this.interval = newInterval;
        this._baseInterval = newInterval;
```

**Review by eonarheim — COMMENTED** [review]
Added some suggestions

File: src/engine/Timer.ts [hunk 2]
```diff
@@ -59,14 +70,37 @@ export class Timer {
...
+        this.interval = interval + new Random().integer(randomRange[0], randomRange[1])
+      }
+      this.randomRange = randomRange;  
+    } else {
+      this.interval = interval;
+    }
 
+    this.repeats = repeats || this.repeats;
+    
     this._callbacks = [];
```
Comment (eonarheim) [inline src/engine/Timer.ts:91]:
```suggestion
    this._callbacks = [];
    this._baseInterval = this.interval = interval;
    if (!!randomRange){
      if(randomRange[0] > randomRange[1]){
        throw new Error('min value must be lower than max value for range');
      }
      //We use the instance of ex.Random to generate the range
      this.random = random ?? new Random();
      this.randomRange = randomRange;

      this.interval = this._generateRandomInterval();
      this.on(() => {
        this.interval = this._generateRandomInterval();
      });
    }
```

File: src/engine/Timer.ts [hunk 3]
```diff
@@ -28,27 +33,33 @@ export class Timer {
   public interval: number = 10;
   public repeats: boolean = false;
   public maxNumberOfRepeats: number = -1;
-
+  public randomRange: [number, number] = [0,0];
+  public random: ex.Random;
```
Comment (eonarheim) [inline src/engine/Timer.ts:37]:
```suggestion
  public random: ex.Random;
  private _baseInterval = 10;
  private _generateRandomInterval = () => {
    return this._baseInterval + this.random.integer(this.randomRange[0], this.randomRange[1])
  }
```

File: src/engine/Timer.ts [hunk 4]
```diff
@@ -59,14 +70,37 @@ export class Timer {
...
       this.on(fcn);
     }
+    if(!!randomRange){
+      if(!!random){
+        this.on(()=> this.interval = interval + this.random.integer(this.randomRange[0], this.randomRange[1]))
+      } else {
+        this.on(()=> this.interval = interval + new Random().integer(this.randomRange[0], this.randomRange[1]))
+      }
+
+    }
```
Comment (eonarheim) [inline src/engine/Timer.ts:103]:
```suggestion

```

Comment (eonarheim) [reply src/engine/Timer.ts:103]: This gets simplified and folded into the above suggestion

File: src/engine/Timer.ts (hunk 2 above)
Comment (eonarheim) [reply src/engine/Timer.ts:91]: We can take advantage of `this.random = random ?? new Random()` (this means if random is not passed in use a new Random) and the `_generateRandomInterval()` helper to simplify a bit

**Comment by malitherl** [comment]
Everything is working on my end too, and all tests are passing. Thank you again for the suggestions and help @eonarheim

**Pull request merged**